drivers/flash/flash_util.c

```diff
@@ -80,3 +87,199 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
 	return -ENOSYS;
 #endif
 }
+
+static bool off_add_overflow(off_t offset, off_t size, off_t *result)
+{
+	if (sizeof(off_t) == sizeof(uint32_t)) {
+		uint32_t end;
+
+		/* account for signedness of off_t due to lack of s32_add_overflow() */
+		if (u32_add_overflow((uint32_t)offset, (uint32_t)size, &end) ||
+		    (end > (UINT32_MAX >> 1))) {
+			return true;
+		}
+	} else if (sizeof(off_t) == sizeof(uint64_t)) {
+		uint64_t end;
+
+		/* account for signedness of off_t due to lack of s64_add_overflow() */
+		if (u64_add_overflow((uint64_t)offset, (uint64_t)size, &end) ||
+		    (end > (UINT64_MAX >> 1))) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+int flash_regions_overlap(off_t a_start, off_t a_size, off_t b_start, off_t b_size)
+{
+	off_t a_end = 0;
+	off_t b_end = 0;
+
+	BUILD_ASSERT((sizeof(off_t) == sizeof(uint32_t)) || (sizeof(off_t) == sizeof(uint64_t)));
+
+	if ((a_start < 0) || (a_size < 0) || (b_start < 0) || (b_size < 0)) {
+		return -EINVAL;
+	}
+
+	if (off_add_overflow(a_start, a_size, &a_end) ||
+	    off_add_overflow(b_start, b_size, &b_end)) {
+		return -EOVERFLOW;
+	}
+
+	return (a_start < b_end) && (a_end > b_start);
+}
+
+int flash_region_crc(const struct device *dev, off_t offset, off_t size, uint32_t *crc,
+		     uint32_t *buf, size_t buf_size)
+{
+	off_t end;
+	uint32_t _buf[4];
+
+	if (!IS_ENABLED(CONFIG_CRC)) {
```

Comment:
would need an #ifdef here otherwise calling `crc32_ieee_update` will give undefined function

Reply:
I'll double check that, but I don't believe that to be the case.

Reply:
Fixed (crc is removed)